checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -52,95 +94,621 @@ public class MustCallInferenceLogic {
+    enMethodTree = ((UnderlyingAST.CFGMethod) cfg.getUnderlyingAST()).getMethod();
+    enMethodElt = TreeUtils.elementFromDeclaration(enMethodTree);
   }
 
   /**
-   * Runs the inference algorithm on the contents of the {@link #cfg} field.
+   * Runs the inference algorithm on the contents of the {@link #cfg} field and parameters.
    *
-   * <p>Operationally, it checks method invocations for fields with non-empty @MustCall obligations
-   * along all paths to the regular exit point in the method body of the method represented by
-   * {@link #cfg}, and updates the {@link #owningFields} set if it discovers an owning field whose
-   * must-call obligations were satisfied along one of the checked paths.
+   * <p>Operationally, it checks method invocations for fields and parameters with
+   * non-empty @MustCall obligations along all paths to the regular exit point in the method body of
+   * the method represented by {@link #cfg}, and updates the {@link #owningFieldToECM} set or
+   * adds @Owning to the formal parameter if it discovers their must-call obligations were satisfied
+   * along one of the checked paths.
    */
   /*package-private*/ void runInference() {
-    Set<Block> visited = new HashSet<>();
-    Deque<Block> worklist = new ArrayDeque<>();
-    Block entry = this.cfg.getEntryBlock();
+
+    Set<BlockWithObligations> visited = new HashSet<>();
+
+    Deque<BlockWithObligations> worklist = new ArrayDeque<>();
+
+    BlockWithObligations entry =
+        new BlockWithObligations(cfg.getEntryBlock(), getNonEmptyMCParams(cfg));
 
     worklist.add(entry);
     visited.add(entry);
 
     while (!worklist.isEmpty()) {
-      Block current = worklist.remove();
+      BlockWithObligations current = worklist.remove();
 
-      for (Node node : current.getNodes()) {
-        if (node instanceof MethodInvocationNode) {
-          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+      Set<Obligation> obligations = new LinkedHashSet<>(current.obligations);
+
+      for (Node node : current.block.getNodes()) {
+        if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+          updateMethodInvocationOrObjectCreationNode(obligations, node);
+          if (node instanceof MethodInvocationNode) {
+            checkMethodInvocation(obligations, (MethodInvocationNode) node);
+          }
+        } else if (node instanceof AssignmentNode) {
+          updateObligationsForAssignment(obligations, (AssignmentNode) node);
         }
       }
 
-      propagateRegPaths(current, visited, worklist);
+      propagateRegPaths(obligations, current.block, visited, worklist);
     }
   }
 
   /**
-   * If the receiver of {@code mNode} is a candidate owning field and the method invocation
-   * satisfies the field's must-call obligation, then adds that field to the {@link #owningFields}
-   * set.
+   * Returns a set of obligations representing the non-empty MustCall parameters of the current
```

Comment:
The use of "current method" here and elsewhere suggests that the current method can change.  I think you mean the method being analyzed by `this`.  Please update throughout.

Reply:
Mike's comment here still seems applicable: the docs here still refer to the "current method", suggesting that its possible for it to change.

Can a single instance of `MustCallInferenceLogic` ever be re-used between different methods? If not, this needs to change (Mike's suggestion is good). If so, the class Javadoc needs to be rewritten.

Reply:
It can't be re-used between different method, it's updated.